syncd/syncd_notifications.cpp

```diff
@@ -185,61 +178,293 @@ void on_port_event(
+    }
+    else
+    {
+        SWSS_LOG_ERROR("unknow notification: %s", notification.c_str());
+    }
+}
+
+// condition variable will be used to notify processing thread
+// that some notiffication arrived
+
+std::condition_variable cv;
+std::mutex queue_mutex;
+std::queue<swss::KeyOpFieldsValuesTuple> ntf_queue;
+
+void enqueue_notification(
+        _In_ std::string op,
+        _In_ std::string data,
+        _In_ std::vector<swss::FieldValueTuple> &entry)
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_INFO("%s %s", op.c_str(), data.c_str());
+
+    swss::KeyOpFieldsValuesTuple item(op, data, entry);
+
+    // this is notification context, so we need to protect queue
+
+    std::lock_guard<std::mutex> lock(queue_mutex);
+
+    ntf_queue.push(item);
+
+    cv.notify_all();
+}
+
+void enqueue_notification(
+        _In_ std::string op,
+        _In_ std::string data)
+{
+    SWSS_LOG_ENTER();
+
+    std::vector<swss::FieldValueTuple> entry;
+
+    enqueue_notification(op, data, entry);
+}
+
+void on_switch_state_change(
+        _In_ sai_switch_oper_status_t switch_oper_status)
+{
+    SWSS_LOG_ENTER();
+
+    std::string s = sai_serialize_switch_oper_status(switch_oper_status);
+
+    enqueue_notification("switch_state_change", s);
+}
+
+void on_fdb_event(
+        _In_ uint32_t count,
+        _In_ sai_fdb_event_notification_data_t *data)
+{
+    SWSS_LOG_ENTER();
+
+    std::string s = sai_serialize_fdb_event_ntf(count, data);
+
+    enqueue_notification("fdb_event", s);
+}
+
+void on_port_state_change(
+        _In_ uint32_t count,
+        _In_ sai_port_oper_status_notification_t *data)
+{
+    SWSS_LOG_ENTER();
+
+    std::string s = sai_serialize_port_oper_status_ntf(count, data);
+
+    enqueue_notification("port_state_change", s);
+}
+
+void on_port_event(
+        _In_ uint32_t count,
+        _In_ sai_port_event_notification_t *data)
+{
+    SWSS_LOG_ENTER();
+
+    std::string s = sai_serialize_port_event_ntf(count, data);
+
+    enqueue_notification("port_event", s);
+}
+
+void on_switch_shutdown_request()
+{
+    SWSS_LOG_ENTER();
+
+    enqueue_notification("switch_shutdown_request", "");
+}
+
 void on_packet_event(
         _In_ const void *buffer,
         _In_ sai_size_t buffer_size,
         _In_ uint32_t attr_count,
         _In_ const sai_attribute_t *attr_list)
 {
-    std::lock_guard<std::mutex> lock(g_ntf_mutex);
-
     SWSS_LOG_ENTER();
 
     SWSS_LOG_ERROR("not implemented");
+}
+
+// determine wheter notification thread is running
+
+volatile bool runThread;
+
+std::mutex ntf_mutex;
+std::unique_lock<std::mutex> ulock(ntf_mutex);
 
-    /*
-       std::string s;
-       sai_serialize_primitive(buffer_size, s);
+bool tryDequeue(
+        _Out_ swss::KeyOpFieldsValuesTuple &item)
+{
+    std::lock_guard<std::mutex> lock(queue_mutex);
 
-       sai_serialize_buffer(buffer, buffer_size, s);
+    SWSS_LOG_ENTER();
 
-       std::vector<swss::FieldValueTuple> entry;
+    if (ntf_queue.size() == 0)
```

Comment:
I saw Qi commenting on another review, asking to use <queue>.empty instead of testing size() ==0, is this an agreed or preferred method? If yes, please change.

Thanks,
Ying

Reply:
done